src/main/java/com/iexec/sms/secret/web2/Web2Secrets.java

```diff
@@ -43,13 +41,41 @@ public class Web2Secrets {
     private List<Secret> secrets;
 
     Web2Secrets(String ownerAddress) {
-        this.ownerAddress = ownerAddress;
-        this.secrets = new ArrayList<>();
+        this(ownerAddress, List.of());
+    }
+
+    public Web2Secrets(String ownerAddress, List<Secret> secrets) {
+        this.ownerAddress = ownerAddress.toLowerCase();
+        this.secrets = Collections.unmodifiableList(secrets);
     }
 
     public Optional<Secret> getSecret(String secretAddress) {
         return secrets.stream()
                 .filter(secret -> secret.getAddress().equals(secretAddress))
                 .findFirst();
     }
+
+    /**
+     * Copies the current {@link Web2Secrets} object, while adding given {@link Secret} to the secrets list.
+     * In case of an update,
+     * given secret is removed from the secrets list before the new object is created.
+     *
+     * @param newSecret {@link Secret} to add to the list
+     * @return A new {@link Web2Secrets} object with given new {@link Secret}.
+     */
+    public Web2Secrets withNewSecret(Secret newSecret) {
```

Comment:
I think I would have prefer 2 `addNewSecret` and `updateExistingSecret` methods.

Here things become a little bit tangled together.
For web2 secrets, we try to ensure that `Web2SecretsService#addSecret` will only add if it can and `Web2SecretsService#updateSecret` will only update if it can.

This produces a situation with double checks where some checks were already performed in Web2SecretsService.
I think that we can achieve a lighter and more straightforward implementation.

Reply:
Please see https://github.com/iExecBlockchainComputing/iexec-sms/pull/134/commits/667228b3b82fc77f7397b1e76062722b82e8798e.